README.md
openmediavault-remoteshare
==========================


debian/changelog
openmediavault-remoteshare (1.2) stable; urgency=low

  * Changed dependency of OMV Package to = 3.0.13 to avoid accidental
    user updates

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Tue, 05 Apr 2016 18:49:58 -0400

openmediavault-remoteshare (1.1) stable; urgency=low

  * Fix Github issue #9
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Tue, 14 Apr 2015 07:00:41 -0500

openmediavault-remoteshare (1.0.8) stable; urgency=low

  * Fix permissions of if-up script

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 28 Dec 2014 09:14:04 -0600

openmediavault-remoteshare (1.0.7) stable; urgency=low

  * Fix ftpfs fstab line
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sat, 27 Dec 2014 07:36:56 -0600

openmediavault-remoteshare (1.0.6) stable; urgency=low

  * Add curlftpfs tab - github issue #5 - Add support for FTP shares
  * Fix unmount - github issue #4 - Share should be unmounted when disabled 
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Thu, 11 Dec 2014 08:25:20 -0600

openmediavault-remoteshare (1.0.5) stable; urgency=low

  * Fix github issue #3 - Spaces should be escaped in Samba share name
  * Add glusterfs tab
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 19 Oct 2014 10:53:51 -0500

openmediavault-remoteshare (1.0.4) stable; urgency=low

  * Bug fixes

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Thu, 18 Sep 2014 12:44:17 -0500

openmediavault-remoteshare (1.0.3) stable; urgency=low

  * Fix bug

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 15 Sep 2014 05:47:27 -0500

openmediavault-remoteshare (1.0.2) stable; urgency=low

  * Minor tweaks

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Tue, 19 Aug 2014 18:22:39 -0500

openmediavault-remoteshare (1.0.1) stable; urgency=low

  * Change dependency

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 08 Aug 2014 14:38:47 -0500

openmediavault-remoteshare (1.0) stable; urgency=low

  * Code formatting
  * add _netdev to smb and nfs shares
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 08 Aug 2014 13:07:00 -0500

openmediavault-remoteshare (0.6.4) stable; urgency=low

  * Add possibility to enable/disable shares for Nfs and Samba
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Wed, 28 May 2014 10:41:09 -0500

openmediavault-remoteshare (0.6.3) stable; urgency=low

  * Add webdav support
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Tue, 27 May 2014 18:33:01 -0500

openmediavault-remoteshare (0.6.2) stable; urgency=low

  * Fix nfs share bug
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 21 Apr 2014 07:10:34 -0500

openmediavault-remoteshare (0.6.1) stable; urgency=low

  * Change fstab insertion point

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Tue, 18 Mar 2014 10:14:27 -0500

openmediavault-remoteshare (0.6) stable; urgency=low

  * Initial release for OMV 0.6

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Tue, 18 Mar 2014 06:30:27 -0500


Sshfs.js
/**
 * @license   http://www.gnu.org/licenses/gpl.html GPL Version 3
 * @author    Volker Theile <volker.theile@openmediavault.org>
 * @author    OpenMediaVault Plugin Developers <plugins@omv-extras.org>
 * @copyright Copyright (c) 2009-2013 Volker Theile
 * @copyright Copyright (c) 2013-2014 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
// require("js/omv/WorkspaceManager.js")
// require("js/omv/workspace/grid/Panel.js")
// require("js/omv/workspace/window/Form.js")
// require("js/omv/workspace/window/plugin/ConfigObject.js")
// require("js/omv/Rpc.js")
// require("js/omv/data/Store.js")
// require("js/omv/data/Model.js")
// require("js/omv/data/proxy/Rpc.js")
// require("js/omv/form/field/SharedFolderComboBox.js")
// require("js/omv/form/field/UserComboBox.js")

Ext.define("OMV.module.admin.service.remoteshare.SshShare", {
    extend   : "OMV.workspace.window.Form",
    requires : [
        "OMV.form.field.SharedFolderComboBox",
        "OMV.form.field.UserComboBox",
    ],
    uses     : [
        "OMV.workspace.window.plugin.ConfigObject"
    ],

    rpcService   : "remoteshare",
    rpcGetMethod : "getSshShare",
    rpcSetMethod : "setSshShare",

    plugins      : [{
        ptype : "configobject"
    }],

    getFormItems : function() {
        var me = this;
        return [{
            xtype      : "checkbox",
            name       : "enable",
            fieldLabel : _("Enable"),
            checked    : true
        },{
            xtype      : "textfield",
            name       : "server",
            fieldLabel : _("Remote Host"),
            allowBlank : false,
            plugins    : [{
                ptype : "fieldinfo",
                text  : _("Hostname, IP address, or fully qualified domain name of the server.")
            }]
        },{
            xtype      : "textfield",
            name       : "remotepath",
            fieldLabel : _("Remote Path"),
            allowBlank : false
        },{
            xtype         : "numberfield",
            name          : "port",
            fieldLabel    : _("Port"),
            vtype         : "port",
            minValue      : 1,
            maxValue      : 65535,
            allowDecimals : false,
            allowBlank    : false,
            value         : 22
        },{
            xtype      : "sharedfoldercombo",
            name       : "sharedfolderref",
            fieldLabel : _("Shared Folder"),
            plugins    : [{
                ptype : "fieldinfo",
                text  : _("Remote path will be mounted as this shared folder for use in other plugins.")
            }]
        },{
            xtype      : "usercombo",
            name       : "username",
            fieldLabel : _("User"),
            value      : "nobody",
            userType   : "normal"
        }];
    }
});

Ext.define("OMV.module.admin.service.remoteshare.SshShares", {
    extend   : "OMV.workspace.grid.Panel",
    requires : [
        "OMV.Rpc",
        "OMV.data.Store",
        "OMV.data.Model",
        "OMV.data.proxy.Rpc"
    ],
    uses     : [
        "OMV.module.admin.service.remoteshare.SshShare"
    ],

    hidePagingToolbar : false,
    stateful          : true,
    stateId           : "1649057b-b1c0-1c48-a4c1-8c8d1fe52c7b",
    columns           : [{
        xtype     : "booleaniconcolumn",
        text      : _("Enabled"),
        sortable  : true,
        dataIndex : "enable",
        stateId   : "enable",
        align     : "center",
        width     : 80,
        resizable : false,
        trueIcon  : "switch_on.png",
        falseIcon : "switch_off.png"
    },{
        text      : _("Remote Host"),
        sortable  : true,
        dataIndex : "server",
        stateId   : "server"
    },{
        text      : _("Remote Path"),
        sortable  : true,
        dataIndex : "remotepath",
        stateId   : "remotepath"
    },{
        text      : _("Port"),
        sortable  : true,
        dataIndex : "port",
        stateId   : "port"
    },{
        text      : _("Shared Folder"),
        sortable  : true,
        dataIndex : "sharename",
        stateId   : "sharename"
    },{
        text      : _("Mount as"),
        sortable  : true,
        dataIndex : "username",
        stateId   : "username"
    }],

    initComponent : function() {
        var me = this;
        Ext.apply(me, {
            store : Ext.create("OMV.data.Store", {
                autoLoad : true,
                model    : OMV.data.Model.createImplicit({
                    idProperty  : "uuid",
                    fields      : [
                        { name : "uuid", type: "string" },
                        { name : "enable", type: "boolean" },
                        { name : "server", type: "string" },
                        { name : "remotepath", type: "string" },
                        { name : "port", type: "integer" },
                        { name : "sharename", type: "string" },
                        { name : "username", type: "string" }
                    ]
                }),
                proxy : {
                    type    : "rpc",
                    rpcData : {
                        service : "remoteshare",
                        method  : "getSshShareList"
                    }
                }
            })
        });
        me.callParent(arguments);
    },

    onAddButton : function() {
        var me = this;
        Ext.create("OMV.module.admin.service.remoteshare.SshShare", {
            title     : _("Add share"),
            uuid      : OMV.UUID_UNDEFINED,
            listeners : {
                scope  : me,
                submit : function() {
                    this.doReload();
                }
            }
        }).show();
    },

    onEditButton : function() {
        var me = this;
        var record = me.getSelected();
        Ext.create("OMV.module.admin.service.remoteshare.SshShare", {
            title     : _("Edit share"),
            uuid      : record.get("uuid"),
            listeners : {
                scope  : me,
                submit : function() {
                    this.doReload();
                }
            }
        }).show();
    },

    doDeletion : function(record) {
        var me = this;
        OMV.Rpc.request({
            scope    : me,
            callback : me.onDeletion,
            rpcData : {
                service : "remoteshare",
                method  : "deleteSshShare",
                params  : {
                    uuid : record.get("uuid")
                }
            }
        });
    }
});

OMV.WorkspaceManager.registerPanel({
    id        : "sshshares",
    path      : "/service/remoteshare",
    text      : _("SSHFS"),
    position  : 50,
    className : "OMV.module.admin.service.remoteshare.SshShares"
});


var/www/openmediavault/js/omv/module/admin/service/remoteshare/Nfs.js
/**
 *
 * @license   http://www.gnu.org/licenses/gpl.html GPL Version 3
 * @author    Volker Theile <volker.theile@openmediavault.org>
 * @author    OpenMediaVault Plugin Developers <plugins@omv-extras.org>
 * @copyright Copyright (c) 2009-2013 Volker Theile
 * @copyright Copyright (c) 2013-2015 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
// require("js/omv/WorkspaceManager.js")
// require("js/omv/workspace/grid/Panel.js")
// require("js/omv/workspace/window/Form.js")
// require("js/omv/workspace/window/plugin/ConfigObject.js")
// require("js/omv/Rpc.js")
// require("js/omv/data/Store.js")
// require("js/omv/data/Model.js")
// require("js/omv/data/proxy/Rpc.js")
// require("js/omv/form/field/SharedFolderComboBox.js")

Ext.define("OMV.module.admin.service.remoteshare.NfsShare", {
    extend : "OMV.workspace.window.Form",
    uses   : [
        "OMV.form.field.SharedFolderComboBox",
        "OMV.workspace.window.plugin.ConfigObject"
    ],

    rpcService   : "remoteshare",
    rpcGetMethod : "getNfsShare",
    rpcSetMethod : "setNfsShare",
    plugins      : [{
        ptype : "configobject"
    }],

    width        : 500,

    getFormItems : function() {
        var me = this;
        return [{
            xtype      : "checkbox",
            name       : "enable",
            fieldLabel : _("Enable"),
            checked    : true
        },{
            xtype      : "textfield",
            name       : "server",
            fieldLabel : _("Server"),
            allowBlank : false,
            plugins    : [{
                ptype : "fieldinfo",
                text  : _("Hostname, IP address, or fully qualified domain name of the server exporting the file system.")
            }]
        },{
            xtype      : "textfield",
            name       : "export",
            fieldLabel : _("Export"),
            allowBlank : false,
            plugins    : [{
                ptype : "fieldinfo",
                text  : _("Path to the exported directory.")
            }]
        },{
            xtype      : "sharedfoldercombo",
            name       : "sharedfolderref",
            fieldLabel : _("Shared Folder"),
            plugins    : [{
                ptype : "fieldinfo",
                text  : _("NFS share will be mounted as this shared folder for use in other plugins.")
            }]
        },{
            xtype      : "checkbox",
            name       : "nfs",
            fieldLabel : _("NFSv4"),
            checked    : false,
            plugins    : [{
                ptype : "fieldinfo",
                text  : _("Leave unchecked for NFSv2 or NFSv3 servers. Check for NFSv4 servers.")
            }]
        },{
            xtype      : "textfield",
            name       : "extraoptions",
            fieldLabel : _("Extra options"),
            allowBlank : true,
            value      : "intr"
        }];
    }
});

Ext.define("OMV.module.admin.service.remoteshare.NfsShares", {
    extend   : "OMV.workspace.grid.Panel",
    requires : [
        "OMV.Rpc",
        "OMV.data.Store",
        "OMV.data.Model",
        "OMV.data.proxy.Rpc"
    ],
    uses     : [
        "OMV.module.admin.service.remoteshare.NfsShare"
    ],

    hidePagingToolbar : false,
    stateful          : true,
    stateId           : "1649057b-b1c0-1c48-a4c1-8c8d1fe52d7b",
    columns           : [{
        xtype     : "booleaniconcolumn",
        text      : _("Enabled"),
        sortable  : true,
        dataIndex : "enable",
        stateId   : "enable",
        align     : "center",
        width     : 80,
        resizable : false,
        trueIcon  : "switch_on.png",
        falseIcon : "switch_off.png"
    },{
        text      : _("Server"),
        sortable  : true,
        dataIndex : "server",
        stateId   : "server"
    },{
        text      : _("Export"),
        sortable  : true,
        dataIndex : "export",
        stateId   : "export"
    },{
        text      : _("Shared Folder"),
        sortable  : true,
        dataIndex : "sharename",
        stateId   : "sharename"
    }],

    initComponent : function() {
        var me = this;
        Ext.apply(me, {
            store : Ext.create("OMV.data.Store", {
                autoLoad : true,
                model    : OMV.data.Model.createImplicit({
                    idProperty  : "uuid",
                    fields      : [
                        { name : "uuid", type: "string" },
                        { name : "enable", type: "boolean" },
                        { name : "server", type: "string" },
                        { name : "export", type: "string" },
                        { name : "sharename", type: "string" }
                    ]
                }),
                proxy : {
                    type    : "rpc",
                    rpcData : {
                        service : "remoteshare",
                        method  : "getNfsShareList"
                    }
                }
            })
        });
        me.callParent(arguments);
    },

    onAddButton : function() {
        var me = this;
        Ext.create("OMV.module.admin.service.remoteshare.NfsShare", {
            title     : _("Add share"),
            uuid      : OMV.UUID_UNDEFINED,
            listeners : {
                scope  : me,
                submit : function() {
                    this.doReload();
                }
            }
        }).show();
    },

    onEditButton : function() {
        var me = this;
        var record = me.getSelected();
        Ext.create("OMV.module.admin.service.remoteshare.NfsShare", {
            title     : _("Edit share"),
            uuid      : record.get("uuid"),
            listeners : {
                scope  : me,
                submit : function() {
                    this.doReload();
                }
            }
        }).show();
    },

    doDeletion : function(record) {
        var me = this;
        OMV.Rpc.request({
            scope    : me,
            callback : me.onDeletion,
            rpcData : {
                service : "remoteshare",
                method  : "deleteNfsShare",
                params  : {
                    uuid : record.get("uuid")
                }
            }
        });
    }
});

OMV.WorkspaceManager.registerPanel({
    id        : "nfsshares",
    path      : "/service/remoteshare",
    text      : _("NFS"),
    position  : 10,
    className : "OMV.module.admin.service.remoteshare.NfsShares"
});


var/www/openmediavault/js/omv/module/admin/service/remoteshare/Gluster.js
/**
 * @license   http://www.gnu.org/licenses/gpl.html GPL Version 3
 * @author    Volker Theile <volker.theile@openmediavault.org>
 * @author    OpenMediaVault Plugin Developers <plugins@omv-extras.org>
 * @copyright Copyright (c) 2009-2013 Volker Theile
 * @copyright Copyright (c) 2013-2015 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
// require("js/omv/WorkspaceManager.js")
// require("js/omv/workspace/grid/Panel.js")
// require("js/omv/workspace/window/Form.js")
// require("js/omv/workspace/window/plugin/ConfigObject.js")
// require("js/omv/Rpc.js")
// require("js/omv/data/Store.js")
// require("js/omv/data/Model.js")
// require("js/omv/data/proxy/Rpc.js")
// require("js/omv/form/field/SharedFolderComboBox.js")

Ext.define("OMV.module.admin.service.remoteshare.GlusterShare", {
    extend : "OMV.workspace.window.Form",
    uses   : [
        "OMV.form.field.SharedFolderComboBox",
        "OMV.workspace.window.plugin.ConfigObject",
        "OMV.form.plugin.LinkedFields"
    ],

    rpcService   : "remoteshare",
    rpcGetMethod : "getGlusterShare",
    rpcSetMethod : "setGlusterShare",

    plugins      : [{
        ptype : "configobject"
    }],

    width        : 550,

    getFormItems : function() {
        var me = this;
        return [{
            xtype      : "checkbox",
            name       : "enable",
            fieldLabel : _("Enable"),
            checked    : true
        },{
            xtype      : "textfield",
            name       : "server",
            fieldLabel : _("Server"),
            allowBlank : false,
            plugins    : [{
                ptype : "fieldinfo",
                text  : _("Hostname, IP address, or fully qualified domain name of the server.")
            }]
        },{
            xtype      : "textfield",
            name       : "volname",
            fieldLabel : _("Volume Name"),
            allowBlank : false
        },{
            xtype      : "sharedfoldercombo",
            name       : "sharedfolderref",
            fieldLabel : _("Shared Folder"),
            plugins    : [{
                ptype : "fieldinfo",
                text  : _("Gluster share will be mounted as this shared folder for use in other plugins.")
            }]
        }];
    }
});

Ext.define("OMV.module.admin.service.remoteshare.GlusterShares", {
    extend   : "OMV.workspace.grid.Panel",
    requires : [
        "OMV.Rpc",
        "OMV.data.Store",
        "OMV.data.Model",
        "OMV.data.proxy.Rpc"
    ],
    uses     : [
        "OMV.module.admin.service.remoteshare.GlusterShare"
    ],

    hidePagingToolbar : false,
    stateful          : true,
    stateId           : "1649057b-b1c0-1c48-a4c1-2b8d1fe52d7b",
    columns           : [{
        xtype     : "booleaniconcolumn",
        text      : _("Enabled"),
        sortable  : true,
        dataIndex : "enable",
        stateId   : "enable",
        align     : "center",
        width     : 80,
        resizable : false,
        trueIcon  : "switch_on.png",
        falseIcon : "switch_off.png"
    },{
        text      : _("Server"),
        sortable  : true,
        dataIndex : "server",
        stateId   : "server"
    },{
        text      : _("Volume Name"),
        sortable  : true,
        dataIndex : "volname",
        stateId   : "volname"
    },{
        text      : _("Shared Folder"),
        sortable  : true,
        dataIndex : "sharename",
        stateId   : "sharename"
    }],

    initComponent : function() {
        var me = this;
        Ext.apply(me, {
            store : Ext.create("OMV.data.Store", {
                autoLoad : true,
                model    : OMV.data.Model.createImplicit({
                    idProperty  : "uuid",
                    fields      : [
                        { name : "uuid", type: "string" },
                        { name : "enable", type: "boolean" },
                        { name : "server", type: "string" },
                        { name : "volname", type: "string" },
                        { name : "sharename", type: "string" }
                    ]
                }),
                proxy : {
                    type    : "rpc",
                    rpcData : {
                        service : "remoteshare",
                        method  : "getGlusterShareList"
                    }
                }
            })
        });
        me.callParent(arguments);
    },

    onAddButton : function() {
        var me = this;
        Ext.create("OMV.module.admin.service.remoteshare.GlusterShare", {
            title     : _("Add share"),
            uuid      : OMV.UUID_UNDEFINED,
            listeners : {
                scope  : me,
                submit : function() {
                    this.doReload();
                }
            }
        }).show();
    },

    onEditButton : function() {
        var me = this;
        var record = me.getSelected();
        Ext.create("OMV.module.admin.service.remoteshare.GlusterShare", {
            title     : _("Edit share"),
            uuid      : record.get("uuid"),
            listeners : {
                scope  : me,
                submit : function() {
                    this.doReload();
                }
            }
        }).show();
    },

    doDeletion : function(record) {
        var me = this;
        OMV.Rpc.request({
            scope    : me,
            callback : me.onDeletion,
            rpcData : {
                service : "remoteshare",
                method  : "deleteGlusterShare",
                params  : {
                    uuid : record.get("uuid")
                }
            }
        });
    }
});

OMV.WorkspaceManager.registerPanel({
    id        : "glustershares",
    path      : "/service/remoteshare",
    text      : _("GlusterFS"),
    position  : 40,
    className : "OMV.module.admin.service.remoteshare.GlusterShares"
});


var/www/openmediavault/js/omv/module/admin/service/remoteshare/Webdav.js
/**
 *
 * @license   http://www.gnu.org/licenses/gpl.html GPL Version 3
 * @author    Volker Theile <volker.theile@openmediavault.org>
 * @author    OpenMediaVault Plugin Developers <plugins@omv-extras.org>
 * @copyright Copyright (c) 2009-2013 Volker Theile
 * @copyright Copyright (c) 2013-2015 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
// require("js/omv/WorkspaceManager.js")
// require("js/omv/workspace/grid/Panel.js")
// require("js/omv/workspace/window/Form.js")
// require("js/omv/workspace/window/plugin/ConfigObject.js")
// require("js/omv/Rpc.js")
// require("js/omv/data/Store.js")
// require("js/omv/data/Model.js")
// require("js/omv/data/proxy/Rpc.js")
// require("js/omv/form/field/SharedFolderComboBox.js")

Ext.define("OMV.module.admin.service.remoteshare.WebdavShare", {
    extend : "OMV.workspace.window.Form",
    uses   : [
        "OMV.form.field.SharedFolderComboBox",
        "OMV.workspace.window.plugin.ConfigObject"
    ],

    rpcService   : "remoteshare",
    rpcGetMethod : "getWebdavShare",
    rpcSetMethod : "setWebdavShare",
    plugins      : [{
        ptype : "configobject"
    }],

    width        : 500,

    getFormItems : function() {
        var me = this;
        return [{
            xtype      : "checkbox",
            name       : "enable",
            fieldLabel : _("Enable"),
            checked    : true
        },{
            xtype      : "textfield",
            name       : "resource",
            fieldLabel : "WebDAV URL",
            allowBlank : false,
            plugins    : [{
                ptype : "fieldinfo",
                text  : _("URL of the WebDAV resource.")
            }]
        },{
            xtype      : "sharedfoldercombo",
            name       : "sharedfolderref",
            fieldLabel : _("Shared Folder"),
            plugins    : [{
                ptype : "fieldinfo",
                text  : _("WebDAV resource will be mounted as this shared folder for use in other plugins.")
            }]
        },{
            xtype      : "textfield",
            name       : "username",
            fieldLabel : _("Username"),
            allowBlank : false
        },{
            xtype      : "passwordfield",
            name       : "password",
            fieldLabel : _("Password"),
            allowBlank : false
        },{
            xtype      : "textfield",
            name       : "extraoptions",
            fieldLabel : _("Extra options"),
            allowBlank : true,
            value      : "_netdev"
        },{
            xtype      : "fieldset",
            title      : _("Davfs2 settings"),
            fieldDefaults: {
                labelSeparator: ""
            },
            items: [{
                border : false,
                html   : "<p>" +
                         _("These settings directly influence how the WebDAV resource is mounted as a local filesystem. Please only change the defaults if you know what you are doing.") +
                         "</p>"
           },{
                xtype      : "checkbox",
                name       : "use_locks",
                fieldLabel : "use_locks",
                checked    : false,
                plugins    : [{
                    ptype : "fieldinfo",
                    text  : _("Whether to lock files on the server when they are opened for writing.")
                }]
            },{
                xtype      : "checkbox",
                name       : "gui_optimize",
                fieldLabel : "gui_optimize",
                checked    : true,
                plugins    : [{
                    ptype : "fieldinfo",
                    text  : _("Optimize fetching file infos for all files in a directory vs. just for the file to be opened.")
                }]
            },{
                xtype      : "checkbox",
                name       : "if_match_bug",
                fieldLabel : "if_match_bug",
                checked    : true,
                plugins    : [{
                    ptype : "fieldinfo",
                    text  : _("Use HEAD instead of If-Match and If-None-Match-headers.")
                }]
            },{
                xtype         : "numberfield",
                name          : "cache_size",
                fieldLabel    : "cache_size",
                minValue      : 1,
                maxValue      : 65535,
                allowDecimals : false,
                allowBlank    : false,
                value         : 1,
                plugins    : [{
                    ptype : "fieldinfo",
                    text  : _("The amount of disk space in MiByte that may be used.")
                }]
            },{
                xtype         : "numberfield",
                name          : "table_size",
                fieldLabel    : "table_size",
                minValue      : 1,
                maxValue      : 65535,
                allowDecimals : false,
                allowBlank    : false,
                value         : 4096,
                plugins       : [{
                    ptype : "fieldinfo",
                    text  : _("The number of entries in the hash table for each known file. The value should be a power of 2.")
                }]
            },{
                xtype         : "numberfield",
                name          : "delay_upload",
                fieldLabel    : "delay_upload",
                minValue      : 0,
                maxValue      : 65535,
                allowDecimals : false,
                allowBlank    : false,
                value         : 1,
                plugins       : [{
                    ptype : "fieldinfo",
                    text  : _("Delay in seconds before a file will be uploaded to the server to avoid upload of temporary files.")
                }]
            },{
                xtype         : "numberfield",
                name          : "buf_size",
                fieldLabel    : "buf_size",
                minValue      : 0,
                maxValue      : 65535,
                allowDecimals : false,
                allowBlank    : false,
                value         : 1,
                plugins       : [{
                    ptype : "fieldinfo",
                    text  : _("Size in KiByte of the buffer used to communicate with the kernel file system.")
                }]
            }]
        }];
    }
});

Ext.define("OMV.module.admin.service.remoteshare.WebdavShares", {
    extend   : "OMV.workspace.grid.Panel",
    requires : [
        "OMV.Rpc",
        "OMV.data.Store",
        "OMV.data.Model",
        "OMV.data.proxy.Rpc"
    ],
    uses     : [
        "OMV.module.admin.service.remoteshare.WebdavShare"
    ],

    hidePagingToolbar : false,
    stateful          : true,
    stateId           : "1649057b-b1c0-1c48-a4c1-8c8d1fe52d7b",
    columns           : [{
        xtype     : "booleaniconcolumn",
        text      : _("Enabled"),
        sortable  : true,
        dataIndex : "enable",
        stateId   : "enable",
        align     : "center",
        width     : 80,
        resizable : false,
        trueIcon  : "switch_on.png",
        falseIcon : "switch_off.png"
    },{
        text      : "WebDAV URL",
        sortable  : true,
        dataIndex : "resource",
        stateId   : "resource"
    },{
        text      : _("Shared Folder"),
        sortable  : true,
        dataIndex : "sharename",
        stateId   : "sharename"
    },{
        text      : _("Username"),
        sortable  : true,
        dataIndex : "username",
        stateId   : "username"
    }],

    initComponent : function() {
        var me = this;
        Ext.apply(me, {
            store : Ext.create("OMV.data.Store", {
                autoLoad : true,
                model    : OMV.data.Model.createImplicit({
                    idProperty  : "uuid",
                    fields      : [
                        { name : "uuid", type: "string" },
                        { name : "enable", type: "boolean" },
                        { name : "resource", type: "string" },
                        { name : "sharename", type: "string" },
                        { name : "username", type: "string" }
                    ]
                }),
                proxy : {
                    type    : "rpc",
                    rpcData : {
                        service : "remoteshare",
                        method  : "getWebdavShareList"
                    }
                }
            })
        });
        me.callParent(arguments);
    },

    onAddButton : function() {
        var me = this;
        Ext.create("OMV.module.admin.service.remoteshare.WebdavShare", {
            title     : _("Add share"),
            uuid      : OMV.UUID_UNDEFINED,
            listeners : {
                scope  : me,
                submit : function() {
                    this.doReload();
                }
            }
        }).show();
    },

    onEditButton : function() {
        var me = this;
        var record = me.getSelected();
        Ext.create("OMV.module.admin.service.remoteshare.WebdavShare", {
            title     : _("Edit share"),
            uuid      : record.get("uuid"),
            listeners : {
                scope  : me,
                submit : function() {
                    this.doReload();
                }
            }
        }).show();
    },

    doDeletion : function(record) {
        var me = this;
        OMV.Rpc.request({
            scope    : me,
            callback : me.onDeletion,
            rpcData : {
                service : "remoteshare",
                method  : "deleteWebdavShare",
                params  : {
                    uuid : record.get("uuid")
                }
            }
        });
    }
});

OMV.WorkspaceManager.registerPanel({
    id        : "webdavshares",
    path      : "/service/remoteshare",
    text      : "WebDAV",
    position  : 30,
    className : "OMV.module.admin.service.remoteshare.WebdavShares"
});
